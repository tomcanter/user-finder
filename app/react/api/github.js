import reqwest from 'reqwest';
import parseLinkHeader from 'utils/parseLinkHeader';
import path from 'path';

const API_ROOT_URL = 'https://api.github.com';

function gitHubAPI(url) {
  var req = reqwest({
    url: path.join(API_ROOT_URL, url),
    type: 'json'
  })
    .then(data => {
      return {
        data,
        pagination: parseLinkHeader(req.request.getResponseHeader('Link'))
      }
    });

  return req;
}

export default gitHubAPI;