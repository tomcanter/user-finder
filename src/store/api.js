import GitHub from 'github-api';

const api = new GitHub({
  token: process.env.USER_FINDER_OAUTH,
});

export default api;
