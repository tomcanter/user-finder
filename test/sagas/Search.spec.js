import {
  call,
  put,
} from 'redux-saga/effects';
import api from 'store/api';
import {
  searchUsers,
} from 'store/sagas/Search';

describe('Saga: searchUsers', () => {

  describe('when a cached result exists', () => {
    it('should yield the result and return', () => {
      const generator = searchUsers({search: 'q=tom'});
      const store = {
        search: {
          cache: {
            'q=tom': 'test',
          },
        },
      };
      const selectAction = generator.next().value;
      expect(selectAction.SELECT.selector(store)).toEqual('test');

      const successAction = generator.next({data: 'foo'}).value;
      expect(successAction).toEqual(
        put({data: 'foo', type: 'SEARCH_SUCCESS', query: 'q=tom'})
      );
    });
  });

  describe('when no cached data exists', () => {
    it('should call the API and normalize the data', () => {
      const generator = searchUsers({search: 'q=tom'});
      generator.next();
      const apiCall = generator.next().value;
      expect(apiCall).toEqual(
        call(api.searchUsers, {q: 'tom'})
      );

      const response = {
        pagination: null,
        total_count: 2,
        items: [
          {id: 123, avatar_url: 'foo', login: 'alecrust', type: 'User'},
          {id: 456, avatar_url: 'foo', login: 'tomcanter', type: 'User'},
        ],
      };
      const successAction = generator.next(response).value;
      expect(successAction.PUT).toMatchSnapshot();
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = searchUsers({search: 'q=tom'});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT).toMatchSnapshot();
    });
  });

});
