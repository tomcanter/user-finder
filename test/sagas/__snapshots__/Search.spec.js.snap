// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga: searchUsers when a request to the API fails should call a failure action 1`] = `
Object {
  "error": true,
  "payload": [Error: it went wrong],
  "type": "SEARCH_FAILURE",
}
`;

exports[`Saga: searchUsers when no cached data exists should call the API and normalize the data 1`] = `
Object {
  "payload": Object {
    "entities": Object {
      "users": Object {
        "123": Object {
          "avatar_url": "foo",
          "id": 123,
          "login": "alecrust",
        },
        "456": Object {
          "avatar_url": "foo",
          "id": 456,
          "login": "tomcanter",
        },
      },
    },
    "pagination": null,
    "result": Array [
      123,
      456,
    ],
    "search": "q=tom",
    "totalResults": 2,
  },
  "type": "SEARCH_SUCCESS",
}
`;
